ic-canister/ic-canister-macros/src/derive.rs

```diff
@@ -130,7 +243,7 @@ fn is_state_attr(attribute: &Attribute) -> bool {
 
 fn get_state_type(input_type: &Type) -> &Type {
     let ref_cell = extract_generic("Rc", input_type, input_type);
-    extract_generic("RefCell", &ref_cell, input_type)
+    extract_generic("RefCell", ref_cell, input_type)
 }
 
 fn extract_generic<'a>(type_name: &str, generic_base: &'a Type, input_type: &'a Type) -> &'a Type {
```

Comment:
We can strip out some of the indent levels and remove the expect if we match on the last segment instead of checking if it's empty first.

```rust
fn extract_generic<'a>(type_name: &str, generic_base: &'a Type, input_type: &'a Type) -> &'a Type
    let v = match generic_base {
        Type::Path(v) => v,
        _ => state_type_error(input_type),
    };

    let last = v
        .path
        .segments
        .iter()
        .last();

    let last_segment = match last {
        Some(segment) => segment,
        None => state_type_error(input_type),
    };

    if last_segment.ident != type_name {
        state_type_error(input_type);
    }

    let arg = match &last_segment.arguments {
        PathArguments::AngleBracketed(arg) => arg,
        _ => state_type_error(input_type),
    };

    if arg.args.len() != 1 {
        state_type_error(input_type);
    }

    match &arg.args[0] {
        GenericArgument::Type(t) => t,
        _ => state_type_error(input_type),
    }
}
```